### PR TITLE
Home task

### PR DESCRIPTION
**Pull request by FedenkoMykhailo**

### CODE REVIEW

File: api/src/main/java/com/eleks/academy/pharmagator/dataproviders/impl/PharmacyLiki24DataProviderImpl.java [hunk 1]
```diff
@@ -29,41 +33,52 @@ public PharmacyLiki24DataProviderImpl(@Qualifier("pharmacyLiki24WebClient") WebC
...
     }
 
-    private Stream<MedicineDto> fetchMedicineDto () {
+    private Stream<MedicineDto> fetchMedicineDto() {
         long pageIndex = 1L;
 
-        Liki24MedicinesResponse liki24MedicinesResponse = webClient.get()
-                .retrieve().bodyToMono(Liki24MedicinesResponse.class)
-                .block();
+        BiConsumer<Long, List<Liki24MedicinesResponse>> fillListByMedicineResponse = (page1, medicinesResponseList1) -> {
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/dataproviders/impl/PharmacyLiki24DataProviderImpl.java]: Please, avoid naming like 'page1' (with numbers in the end)

File: api/src/main/java/com/eleks/academy/pharmagator/scheduler/Scheduler.java [hunk 2]
```diff
@@ -40,27 +40,21 @@ private void storeToDatabase(MedicineDto dto) {
...
-
-        } else {
+        existMedicine.ifPresentOrElse(medicine -> {
+            priceRepository
+                    .findByPharmacyIdAndMedicineId(medicine.getId(), dto.getPharmacyId())
+                    .ifPresentOrElse(price -> {
+                        price.setPrice(dto.getPrice());
+                        price.setUpdatedAt(Instant.now());
+                        priceRepository.save(price);
+                    }, () -> priceRepository.save(Price.builder()
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/scheduler/Scheduler.java]: Please, extract builder

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java [hunk 3]
```diff
@@ -45,7 +47,7 @@ public void findAllPharmacies_findIds_ok() throws Exception {
         try {
             DatabaseOperation.REFRESH.execute(this.dataSourceConnection, readDataset());
 
-            this.mockMvc.perform(MockMvcRequestBuilders.get("/pharmacies"))
+            this.mockMvc.perform(MockMvcRequestBuilders.get(URL_TEMPLATE))
```
Comment (Oleksandr-Eleks) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java:50]: Kudos! :)

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@WebMvcTest
+@AutoConfigureMockMvc
+@ContextConfiguration(classes = {PharmacyController.class})
+public class PharmacyControllerTest {
+    public static final String API_PHARMACY_URL = "/pharmacies";
+
+    private ObjectMapper mapper;
+    private Pharmacy pharmacy1;
```
Comment (Oleksandr-Eleks) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java]: I don't like it, but for test this naming is not critical.
Still, I'd prefer to change it.

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        assertEquals(mapper.writeValueAsString(pharmacy1), actual);
+        verify(pharmacyService, times(1)).savePharmacy(Mockito.any(PharmacyDto.class));
+    }
+
+    @Test
+    void testUpdatePharmacy() throws Exception {
+
+        when(pharmacyService.updatePharmacy(any(PharmacyDto.class), Mockito.anyLong())).thenReturn(PharmacyDtoMapper.toDto(pharmacy1));
+
+        String actual = mockMvc.perform(put(API_PHARMACY_URL + "/1/")
```
Comment (Oleksandr-Eleks) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java]: Do not use ID like this. Even here it is better to get this ID from mocked object

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+    }
+
+    @Test
+    void deleteMedicine() throws Exception {
+
+        String expected = "pharmacy was successfully deleted!";
+        doNothing().when(pharmacyService).deletePharmacy(Mockito.anyLong());
+
+        String actual = mockMvc.perform(delete(API_PHARMACY_URL + "/1/")
```
Comment (Oleksandr-Eleks) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java]: The same

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                .medicineId(2021102502L)
+                .externalId("2021102502")
+                .price(BigDecimal.valueOf(100))
+                .updatedAt(Instant.now())
+                .build();
+    }
+
+    @Test
+    void testFindAllPrice() throws Exception {
+        var priceList = List.of(this.price1, price2);
```
Comment (Oleksandr-Eleks) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerTest.java]: Why only one 'this'? =)

File: api/src/test/java/com/eleks/academy/pharmagator/impl/MedicineServiceImplTest.java [hunk 8]
```diff
@@ -1,31 +1,134 @@
...
+    }
+
+    @Test
+    void testFindMedicineByTitle() {
+        Mockito.when(medicineRepository.findByTitle(Mockito.anyString())).thenReturn(Optional.ofNullable(medicine1));
+        var actual = medicineService.findMedicineByTitle(Mockito.anyString());
+
+        assertEquals(medicine1, actual.get());
     }
 }
```
Comment (Oleksandr-Eleks) [inline api/src/test/java/com/eleks/academy/pharmagator/impl/MedicineServiceImplTest.java:132]: Try always to have a blank line in the end of the file, to not trigger GitHub.

File: api/src/test/java/com/eleks/academy/pharmagator/impl/PharmacyServiceImplTest.java [hunk 9]
```diff
@@ -1,31 +1,101 @@
...
+        var expected = Stream.of(pharmacy1).map(PharmacyDtoMapper::toDto).collect(Collectors.toList());
+
+        assertFalse(actual.isEmpty());
+        assertEquals(1, actual.size());
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    void testFindPharmacyByIdWhenIdExists() {
+        Mockito.when(pharmacyRepository.findById(Mockito.anyLong())).thenReturn(java.util.Optional.ofNullable(pharmacy1));
```
Comment (Oleksandr-Eleks) [inline api/src/test/java/com/eleks/academy/pharmagator/impl/PharmacyServiceImplTest.java]: Such test is testing your mock. Here You define behavior in test, and then testing it. It's a bit useless...

File: api/src/test/java/com/eleks/academy/pharmagator/impl/PharmacyServiceImplTest.java [hunk 10]
```diff
@@ -1,31 +1,101 @@
...
+        Mockito.when(pharmacyRepository.findById(Mockito.anyLong())).thenReturn(java.util.Optional.empty());
+        var pharmacyNotFoundException = assertThrows(PharmacyNotFoundException.class, () -> pharmacyService.findPharmacyById(Mockito.anyLong()));
+        String expected = String.format("farmacy not found for id: %s", 0);
+
+        assertEquals(expected, pharmacyNotFoundException.getMessage());
     }
 
     @Test
     void testSavePharmacy() {
+        Mockito.when(pharmacyRepository.save(Mockito.any(Pharmacy.class))).thenReturn(pharmacy1);
```
Comment (Oleksandr-Eleks) [inline api/src/test/java/com/eleks/academy/pharmagator/impl/PharmacyServiceImplTest.java]: The same. It rather should be DBunit test or so...

File: api/src/test/java/com/eleks/academy/pharmagator/impl/MedicineServiceImplTest.java (hunk 8 above)
Comment (FedenkoMykhailo) [reply api/src/test/java/com/eleks/academy/pharmagator/impl/MedicineServiceImplTest.java:132]: Done

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerTest.java (hunk 7 above)
Comment (FedenkoMykhailo) [reply api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerTest.java]: Changed

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java (hunk 6 above)
Comment (FedenkoMykhailo) [reply api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java]: Fixed

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java (hunk 5 above)
Comment (FedenkoMykhailo) [reply api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java]: Fixed

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java (hunk 4 above)
Comment (FedenkoMykhailo) [reply api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerTest.java]: rename variable

File: api/src/main/java/com/eleks/academy/pharmagator/scheduler/Scheduler.java (hunk 2 above)
Comment (FedenkoMykhailo) [reply api/src/main/java/com/eleks/academy/pharmagator/scheduler/Scheduler.java]: Done

File: api/src/main/java/com/eleks/academy/pharmagator/dataproviders/impl/PharmacyLiki24DataProviderImpl.java (hunk 1 above)
Comment (FedenkoMykhailo) [reply api/src/main/java/com/eleks/academy/pharmagator/dataproviders/impl/PharmacyLiki24DataProviderImpl.java]: rename variable

**Pull request closed without merging**